### PR TITLE
Add ActOnCliffordTableauArgs. Support it in _act_on_ for common gates and measurements.

### PR DESCRIPTION
**Pull request by smitsanghavi**
#2948 #2423

### CODE REVIEW

**Comment by smitsanghavi** [comment]
Friendly Ping!

File: cirq/ops/common_gates_test.py [hunk 1]
```diff
@@ -272,6 +261,232 @@ def test_h_str():
     assert str(cirq.H**0.5) == 'H^0.5'
 
 
+def test_x_act_on():
```
Comment (Strilanc) [inline cirq/ops/common_gates_test.py:264]: Create a testing method `def assert_clifford_tableau_effect_matches_unitary():`

Implement the method as follows:

1. Create a circuit that prepares a Bell pair for each qubit to be input into the gate (H + CNOT), then runs the gate (acting on one qubit from each Bell pair).
2. Create a method `stabilizers_to_state_vector` which creates the state stabilized by a given list of commuting stabilizers. Do this by generating a random state and then iteratively projecting it onto each stabilizer's +1 eigenstate.
3. Compare the result of `cirq.final_state_vector(circuit)` to (the equivalent of) `stabilizers_to_state(cirq.final_clifford_tableau(circuit))`.

Alternatively, you can check that the state returned by the state vector simulator is in fact stabilized by each of the claimed stabilizers in the Clifford tableau. That would be more efficient in general.

The key thing is that we want to be able to base the correctness of the Clifford simulation on the correctness of the state vector simulator.

Comment (Strilanc) [reply cirq/ops/common_gates_test.py:264]: Maybe we could implement a utility method `state_vector_has_stabilizer(np.ndarray, PauliString)`.

Comment (smitsanghavi) [reply cirq/ops/common_gates_test.py:264]: Sounds good. Went with the state_vector_has_stabilizer approach.

**Comment by smitsanghavi** [comment]
Tests failing due to #2468

**Comment by smitsanghavi** [comment]
Please review

File: cirq/testing/consistent_act_on.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+from cirq.circuits.circuit import Circuit
+from cirq.devices import LineQubit
+from cirq.ops import common_gates
+from cirq.ops.dense_pauli_string import DensePauliString
+from cirq import protocols
+from cirq.sim import act_on_state_vector_args, final_state_vector
+from cirq.sim.clifford import act_on_clifford_tableau_args, clifford_tableau
+
+
+def state_vector_has_stabilizer(state_vector: np.ndarray,
```
Comment (Strilanc) [inline cirq/testing/consistent_act_on.py:28]: This method currently mutates the `state_vector` argument and makes a copy that it doesn't mutate. Switch it around so the input argument is not mutated but the copy is mutated.

File: cirq/testing/consistent_act_on.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                                stabilizer: DensePauliString) -> bool:
+    original_state_vector = state_vector.copy()
+    args = act_on_state_vector_args.ActOnStateVectorArgs(
+        target_tensor=state_vector,
+        available_buffer=np.empty(state_vector.shape, dtype=np.complex64),
```
Comment (Strilanc) [inline cirq/testing/consistent_act_on.py:33]: Prefer `np.empty_like(state_vector)`.

File: cirq/testing/consistent_act_on.py (hunk 2 above)
Comment (Strilanc) [inline cirq/testing/consistent_act_on.py:28]: Add a docstring. Note that the stabilizer has to be correct even including global phase (e.g. Z is a stabilizer of |0> but not |1>; that would be stabilized by -Z instead).

File: cirq/testing/consistent_act_on.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        circuit.append([
+            common_gates.H(qubits[2 * i]),
+            common_gates.CNOT(qubits[2 * i], qubits[2 * i + 1])
+        ])
+    circuit.append(val.on(*qubits[::2]))
+    state_vector = np.reshape(final_state_vector(circuit, qubit_order=qubits),
+                              protocols.qid_shape(qubits))
+
+    tableau = clifford_tableau.CliffordTableau(len(qubits))
+    for op in circuit.all_operations():
```
Comment (Strilanc) [inline cirq/testing/consistent_act_on.py:65]: Extract this into a local private `final_tableau` method that we can later make globally accessible.

Comment (smitsanghavi) [reply cirq/testing/consistent_act_on.py:96]: Done.

File: cirq/testing/consistent_act_on.py (hunk 2 above)
Comment (smitsanghavi) [reply cirq/testing/consistent_act_on.py:29]: Done.

File: cirq/testing/consistent_act_on.py (hunk 3 above)
Comment (smitsanghavi) [reply cirq/testing/consistent_act_on.py]: Thanks for the suggestion!

File: cirq/testing/consistent_act_on.py (hunk 2 above)
Comment (smitsanghavi) [reply cirq/testing/consistent_act_on.py:29]: Done

**Comment by smitsanghavi** [comment]
> Add the "clifford matches" test to the `_assert_meets_standards_helper` method.

This also helped me find that Operations were not supported. Fixed that.

**Comment by Strilanc** [comment]
Very nice!

**Pull request merged**